unlock-app/src/components/creator/CreatorLockForm.js

```diff
@@ -14,86 +14,157 @@ import {
   LockLabel,
   LockDuration,
   LockKeys,
-  LockRowGrid,
 } from './CreatorLock'
 import { LockStatus } from './lock/CreatorLockStatus'
 import { createLock } from '../../actions/lock'
+import { setError, resetError } from '../../actions/error'
+import {
+  FORM_LOCK_NAME_MISSING,
+  FORM_EXPIRATION_DURATION_INVALID,
+  FORM_MAX_KEYS_INVALID,
+  FORM_KEY_PRICE_INVALID,
+} from '../../errors'
 
 export class CreatorLockForm extends React.Component {
   constructor(props, context) {
     super(props, context)
     this.state = {
-      expirationDuration: 30,
-      expirationDurationUnit: 86400, // Days
-      keyPrice: '0.01',
-      keyPriceCurrency: 'ether',
-      maxNumberOfKeys: 10,
-      unlimitedKeys: false,
-      name: 'New Lock',
+      expirationDuration: props.expirationDuration,
+      expirationDurationUnit: props.expirationDurationUnit, // Days
+      keyPrice: props.keyPrice,
+      keyPriceCurrency: props.keyPriceCurrency,
+      maxNumberOfKeys: props.maxNumberOfKeys,
+      unlimitedKeys: props.unlimitedKeys,
+      name: props.name,
     }
+    const { validityState: valid, errors } = this.valid(this.state)
+    this.state.valid = valid
+    this.state.errors = errors
     this.handleSubmit = this.handleSubmit.bind(this)
     this.handleCancel = this.handleCancel.bind(this)
     this.handleChange = this.handleChange.bind(this)
     this.handleUnlimitedClick = this.handleUnlimitedClick.bind(this)
-    this.maxNumberOfKeysRef = React.createRef()
+  }
+
+  valid(state) {
+    // the list of errors we will pass to setError
+    const errors = []
+    // for each field, retrieve the error triggered by current state
+    // and then make sure we set it as existing.
+    const validityState = [
+      'expirationDuration',
+      //'expirationDurationUnit',
+      'keyPrice',
+      //'keyPriceCurrency',
+      'maxNumberOfKeys',
+      'name',
+    ].reduce((fieldValidity, field) => {
+      // invalidError will either be the error name or false
+      const invalidError = this.validate(field, state[field])
+      fieldValidity[field] = !invalidError
+      if (!invalidError) {
+        return fieldValidity
+      }
+      errors.push(invalidError)
+      return fieldValidity
+    }, {})
+
+    // the form can be submitted if and only if there are no errors triggered by any field
+    validityState.formValid = errors.length === 0
+    return { validityState, errors }
   }
 
   validate(name, value) {
-    let { unlimitedKeys } = this.state
-    return (
-      (name === 'name' && typeof value === 'string' && value.length > 0) ||
-      ((name === 'expirationDuration' || name === 'maxNumberOfKeys') &&
-        !isNaN(value) &&
-        value > 0) ||
-      (name === 'maxNumberOfKeys' && value === '∞' && unlimitedKeys) ||
-      (name === 'keyPrice' && !isNaN(value) && value >= 0)
-    )
+    const missing = val => val === undefined || val === null || val === ''
```

Comment:
```suggestion
    const missing = val => (val === undefined || val === null || val === '')
```

Reply:
We can probably extract all of these functions from this one. Either at the top of the file or (even better), in helper file (`validations.js`?) which can be done in its own PR.